### PR TITLE
Clear debug information for kill and replacement

### PR DESCRIPTION
**Pull request by jaebaek**
For many spirv-opt passes such as simplify-instructions pass, we have to
correctly clear the OpenCL.DebugInfo.100 debug information for
KillInst() and ReplaceAllUses(). If we keep some debug information that
disappeared because of KillInst() and ReplaceAllUses(), adding new
DebugValue instructions will generate incorrect information. This CL
update DebugInfoManager and IRContext to correctly clear debug
information.

### CODE REVIEW

**Comment by jaebaek** [comment]
@greg-lunarg I think you need this PR for the debug information preservation in ADCE. Could you please review the PR if it has some issues with ADCE?

File: source/opt/debug_info_manager.h [hunk 1]
```diff
@@ -157,17 +163,18 @@ class DebugInfoManager {
   // in |inst| must not already be registered.
   void RegisterDbgFunction(Instruction* inst);
 
-  // Register the DebugDeclare instruction |dbg_declare| into
-  // |var_id_to_dbg_decl_| using OpVariable id |var_id| as a key.
-  void RegisterDbgDeclare(uint32_t var_id, Instruction* dbg_declare);
+  // Register the DebugDeclare or DebugValue instruction |dbg_inst| into
+  // |var_id_to_dbg_decl_| using |id| as a key.
```
Comment (s-perron) [inline source/opt/debug_info_manager.h:167]: |id| is not a variable.  I think you meant |var_or_value_id|.

File: source/opt/debug_info_manager.h [hunk 2]
```diff
@@ -191,9 +198,9 @@ class DebugInfoManager {
   // operand is the function.
   std::unordered_map<uint32_t, Instruction*> fn_id_to_dbg_fn_;
 
-  // Mapping from local variable ids to DebugDeclare instructions whose
-  // operand is the local variable.
-  std::unordered_map<uint32_t, std::vector<Instruction*>> var_id_to_dbg_decl_;
+  // Mapping from variable or value ids to DebugDeclare or DebugValue
+  // instructions whose operand is the variable or value.
+  std::unordered_map<uint32_t, std::list<Instruction*>> var_id_to_dbg_decl_;
```
Comment (s-perron) [inline source/opt/debug_info_manager.h:203]: Why did you change to a list?

File: source/opt/ir_context.cpp [hunk 3]
```diff
@@ -424,10 +430,12 @@ void IRContext::KillOperandFromDebugInstructions(Instruction* inst) {
...
-  // Notice that we do not need anythings to do for local variables.
-  // DebugLocalVariable does not have an OpVariable operand. Instead,
-  // DebugDeclare/DebugValue has an OpVariable operand for a local
-  // variable. The function inlining pass handles it properly.
+  // Kill DebugValue or DebugDeclare whose operand is |id|.
+  std::vector<Instruction*> dbg_decl_or_value;
+  get_debug_info_mgr()->ForEachDebugDeclareOrValue(
+      id,
+      [&dbg_decl_or_value](Instruction* i) { dbg_decl_or_value.push_back(i); });
+  for (auto* i : dbg_decl_or_value) KillInst(i);
```
Comment (s-perron) [inline source/opt/ir_context.cpp:438]: I'm not sure about this.  If the DebugDeclare or DebugValue are still in the code when an instruction is killed, then it is likely that the optimization is wrong.  The value should have been replaced or the debug instruction explicitly removed.

Comment (s-perron) [reply source/opt/ir_context.cpp:438]: Also, don't add any type of assert.  We want to allow the passes to delete instructions in any order the want.  An error will be picked by by the validator it the pass does not delete the instruction.

File: source/opt/debug_info_manager.h (hunk 2 above)
Comment (s-perron) [reply source/opt/debug_info_manager.h]: unorder_set is fine.  Avoid std::list, it is known to have poor implementations.  In general, if you know things will be small, you can never go wrong with std::vector.  It has so much less overhead than the other containers, that in practice it can still be faster.

>  std::vector::erase calls the destructor and the constructor O(N) times

A pointer does not have a constructor or destructor, so it will not be called.

File: test/opt/local_ssa_elim_test.cpp [hunk 4]
```diff
@@ -2185,13 +2185,13 @@ TEST_F(LocalSSAElimTest, DebugValueForReferenceVariable) {
...
+; CHECK: OpExtInst %void [[ext:%\d+]] DebugLocalVariable [[f_name]]
 ; CHECK: [[dbg_i:%\w+]] = OpExtInst %void [[ext]] DebugLocalVariable [[i_name]]
-; CHECK: [[dbg_x:%\w+]] = OpExtInst %void [[ext]] DebugLocalVariable [[x_name]]
+; CHECK: OpExtInst %void [[ext]] DebugLocalVariable [[x_name]]
 
 ; CHECK:      OpStore %f %float_0
-; CHECK-NEXT: OpExtInst %void [[ext]] DebugValue [[dbg_f]] %float_0
-; CHECK-NEXT: OpExtInst %void [[ext]] DebugValue [[dbg_x]] %float_0
+; CHECK-NEXT: OpExtInst %void [[ext]] DebugValue {{%\w+}} %float_0
+; CHECK-NEXT: OpExtInst %void [[ext]] DebugValue {{%\w+}} %float_0
```
Comment (s-perron) [inline test/opt/local_ssa_elim_test.cpp:2194]: Why are you not checking the local variables anymore?  If the issue is order, you can use CHECK-DAG.

Comment (jaebaek) [reply test/opt/local_ssa_elim_test.cpp]: Thanks. CHECK-DAG is what I needed.

File: source/opt/debug_info_manager.h (hunk 2 above)
Comment (jaebaek) [reply source/opt/debug_info_manager.h]: I see. Thank you for the explanation!
